demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+        if (ERROR_SUCCESS != status) {
+            throw std::runtime_error("PdhGetFormattedCounterValue() failed");
+        }
+
+        idleNonIdleCpuStat.emplace_back(DisplayValue.doubleValue, 1 - DisplayValue.doubleValue);
+    }
+
+    std::vector<double> cpuLoad(idleNonIdleCpuStat.size());
+    for (std::size_t i = 0; i < idleNonIdleCpuStat.size(); ++i) {
+        cpuLoad[i] = static_cast<double>(idleNonIdleCpuStat[i].first);
+    }
+
+    if (meanCpuLoad.empty()) {
+        meanCpuLoad = cpuLoad;
+        samplesNumber = 1;
+    } else {
+        for (std::size_t i = 0; i < meanCpuLoad.size(); ++i) {
+            meanCpuLoad[i] = (meanCpuLoad[i] * samplesNumber + cpuLoad[i]) / (samplesNumber + 1);
+        }
+        ++samplesNumber;
+    }
+    cpuLoadHistory.push_back(std::move(cpuLoad));
+    if (cpuLoadHistory.size() > historySize) {
+        cpuLoadHistory.pop_front();
+    }
+}
+
+std::deque<std::vector<double>> CpuMonitor::getLastHistory() const {
+    return cpuLoadHistory;
+}
+
+std::vector<double> CpuMonitor::getMeanCpuLoad() const {
+    return meanCpuLoad;
+}
+#else
+#include <regex>
+#include <utility>
+#include <fstream>
+
+namespace {
+std::size_t getNCores() {
+    std::ifstream cpuinfo("/proc/cpuinfo");
+    return std::count(std::istream_iterator<std::string>(cpuinfo),
+        std::istream_iterator<std::string>(),
+        std::string("processor"));
+}
+
+std::vector<std::pair<unsigned long, unsigned long>> getIdleNonIdleCpuStat(std::size_t nCores) {
+    std::vector<std::pair<unsigned long, unsigned long>> idleNonIdleCpuStat(nCores);
+    std::ifstream proc_stat("/proc/stat");
+    std::string line;
+    std::smatch match;
+    std::getline(proc_stat, line);
+    while (proc_stat.good())
+    {
+        std::regex core_jiffies("^cpu(\\d+)\\s+"
+            "(\\d+)\\s+" // user
+            "(\\d+)\\s+" // nice
+            "(\\d+)\\s+" // system
+            "(\\d+)\\s+" // idle
+            "(\\d+)\\s+" // iowait
+            "(\\d+)\\s+" // irq
+            "(\\d+)\\s+" // softirq
+            "(\\d+)\\s+" // steal
+            "(\\d+)\\s+" // guest
+            "(\\d+)$");  // guest_nice
+        if (std::regex_match(line, match, core_jiffies))
+        {
+            unsigned long idleInfo = stoul(match[5]) + stoul(match[6]),
+                non_idleinfo = stoul(match[2])
+                    + stoul(match[3])
+                    + stoul(match[4])
+                    + stoul(match[7])
+                    + stoul(match[8])
+                    + stoul(match[9]), // it doesn't handle overflows and overflows of /proc/stat values
+                core_id = stoul(match[1]);
+            idleNonIdleCpuStat[core_id].first = idleInfo;
+            idleNonIdleCpuStat[core_id].second = non_idleinfo;
+        }
+        std::getline(proc_stat, line);
+    }
+    return idleNonIdleCpuStat;
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
```

Comment:
There's a lot of duplication here with the Windows version of the code. You should move the common code out of the `#if`.

Reply:
moved

Reply:
There is more common code than that. :-) Parts of the constructor, `setHistorySize` and `collectData` are all duplicated too.

IMO, what you should do here is expand the `PerformanceCounter` class. Move everything that is OS-dependent into there, then the `CpuMonitor` class itself can become entirely OS-independent.

Reply:
I think further emphasizing of common parts should be avoided because it introduces a number of additional switches between Windows and Linux and I find them annoying and hard for understanding.
Why should I move everything that is OS-dependent into `PerformanceCounter`? Most of `CpuMonitor`'s logic is already OS specific. Moving it into a separate class will introduce another level of nesting because `PerformanceCounter` is supposed to live only for period when the data needs to be collected and `openQuery()` manages this object creation, so I can't just move everything into `PerformanceCounter`.

Reply:
> I think further emphasizing of common parts should be avoided because it introduces a number of additional switches between Windows and Linux 

Does it? In my mind it actually _reduces_ the number of switches. Here's what I imagine it would look like:

* `CpuMonitor::openQuery` is gone (merged into the `PerformanceCounter` constructor).
* `CpuMonitor::closeQuery` is gone (unneeded).
* `CpuMonitor::CpuMonitor` is the same, except the `nCores` member initializer is gone.
* `CpuMonitor::setHistorySize` looks like this:

```c++
    if (0 == historySize && 0 != size) {
        performanceCounter.reset(new PerformanceCounter);
    } else if (0 != historySize && 0 == size) {
        performanceCounter.reset();
    }
    historySize = size;
    std::size_t newSize = std::min(size, cpuLoadHistory.size());
    cpuLoadHistory.erase(cpuLoadHistory.begin(), cpuLoadHistory.end() - newSize);
```

* `CpuMonitor::collectData` looks like this:

```c++
    auto cpuLoad = performanceCounter->query();

    for (std::size_t i = 0; i < cpuLoad.size(); ++i) {
        cpuLoadSum[i] += cpuLoad[i];
    }
    ++samplesNumber;

    cpuLoadHistory.push_back(std::move(cpuLoad));
    if (cpuLoadHistory.size() > historySize) {
        cpuLoadHistory.pop_front();
    }
```

This makes it so that the only `#if` required is around the `PerformanceCounter` class. As a bonus, it also nicely separates the collection of metrics from their aggregation.

Reply:
Updated.